### DISCUSSION TITLE
How to allow http server in android webview

### DISCUSSION THREAD

**Discussion by proneon267** [beeware/toga#1801] [Q&A]
I am trying to allow http traffic in android webview, but it return error clear text traffic is not permitted . So, I added ``android:usesCleartextTraffic="true"`` to ``AndroidManifest.xml``. after I do ``briefcase create android`` and before ``briefcase build android``, but still it returns clear text traffic not permitted. Is there any way to allow http localhost traffic in android webview?

**Comment by freakboy3742** [comment]
The default Briefcase Android template should be sufficient to display content from a localhost http server, without any additional permissions. By way of proof, the `positron-static` and `positron-django` demos both work in my experience. Do these demos work for you? If not, then we have something to investigate; if they do, then it sounds like there's something not quite right with your localhost server.

**Reply by proneon267** [reply]
Hello
I cloned the master repo of toga and tried to run `positron-djanago` from examples folder. But I cannot run or build it. I am getting error: 
```
× Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [23 lines of output]
      Traceback (most recent call last):
        File "/home/proneon267/.pyenv/versions/3.8.16/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 363, in <module>
          main()
        File "/home/proneon267/.pyenv/versions/3.8.16/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 345, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
        File "/home/proneon267/.pyenv/versions/3.8.16/lib/python3.8/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 130, in get_requires_for_build_wheel
          return hook(config_settings)
        File "/tmp/pip-build-env-ivz47fjr/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 338, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=['wheel'])
        File "/tmp/pip-build-env-ivz47fjr/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 320, in _get_build_requires
          self.run_setup()
        File "/tmp/pip-build-env-ivz47fjr/overlay/lib/python3.8/site-packages/setuptools/build_meta.py", line 335, in run_setup
          exec(code, locals())
        File "<string>", line 5, in <module>
        File "/tmp/pip-build-env-ivz47fjr/overlay/lib/python3.8/site-packages/setuptools_scm/__init__.py", line 149, in get_version
          _version_missing(config)
        File "/tmp/pip-build-env-ivz47fjr/overlay/lib/python3.8/site-packages/setuptools_scm/__init__.py", line 109, in _version_missing
          raise LookupError(
      LookupError: setuptools-scm was unable to detect version for /home/proneon267/Downloads/toga-main.
      
      Make sure you're either building from a fully intact git repository or PyPI tarballs. Most other sources (such as GitHub's tarballs, a git checkout without the .git folder) don't contain the necessary metadata and will not work.
      
      For example, if you're using pip, instead of https://github.com/user/proj/archive/master.zip use git+https://github.com/user/proj.git#egg=proj
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.
```
it says to use intact repo. Does cloning not gives an intact repo? 

I did not know there was a separate android template in briefcase. So, how do I generate the default android template with briefcase?

**Reply by freakboy3742** [reply]
The error you've shown suggests that you are missing some of the development requirements needed to run packages (specifically, setuptools_scm); However, it's very hard to diagnose what has gone wrong when you only provide a partial error trace, and don't tell us what you've done to generate that error.

You say you "ran positron-django" - How? With Briefcase? with `python -m`? Is this the result of running on macOS? Linux? Android? Are you using a GitHub clone, or have you obtained the source code some other way? How did you install Toga into your environment (if you've done that at all)?

As for Briefcase - yes, it has an [android template](https://github.com/beeware/briefcase-android-gradle-template). You don't need to do *anything* to get that template - Briefcase manages the process of obtaining and using the template in your Android deployment. You generate the default template by using `briefcase android create`.

**Reply by proneon267** [reply]
Hello
Sorry for missing out the system information:
The host OS is `linux`
Python version is `3.8.16`
I have installed beeware `0.3` and toga `0.3` from pypi. 
So, first I cloned the github repo and then I went into the `positron-django` directory. Then, I ran the command `briefcase dev` and then I got the error that I posted. 
Also, I tried `briefcase create android` and `briefcase build android` but got the same error. 
I will try to install `setuptools_scm` and rerun `briefcase dev` to check if it works
If you need, I can provide the complete error trace.

**Reply by mhsmith** [reply]
> I have installed beeware `0.3` and toga `0.3` from pypi.

Are you sure you're on the newest version of Briefcase? Installing the `beeware` package won't automatically update it. Try this:
* Run `briefcase --version`
* If it's not the newest version from [this list](https://pypi.org/project/briefcase/#history), run `pip install -U briefcase`.
* Check your `android` directory and make sure it contains [these lines](https://github.com/beeware/briefcase-android-gradle-template/pull/53/files). If it doesn't, regenerate the directory by running `briefcase create android`.

**Reply by proneon267** [reply]
Good news all works now!
I ran `briefcase --version` and I get `0.3.12`

I checked the `AndroidManifest.xml` and the `network_security_config.xml` files and the lines are present. 

But in `network_security_config.xml`, there is only one domain `<domain includeSubdomains="true">127.0.0.1</domain>` and I was using `localhost`, so I was getting the clear text error. Is there any security reason for not also including `localhost`?

**Reply by mhsmith** [reply]
Strange, I [specifically tested](https://github.com/beeware/briefcase-android-gradle-template/pull/53#pullrequestreview-1142084767) with `localhost` when we made this change, and it worked fine. Maybe it varies from device to device.

**Reply by proneon267** [reply]
Hello
On which android version did you test? I tested the following cases on an Android 10 physical device and also on Windows Subsystem For Android:
| Value in `network_security_config.xml` | Value in `app.py` | HTTP Traffic Permission |
|------------------------------------------------------|-------------------------|----------------------|
|127.0.0.1                                                       | 127.0.0.1              | Allowed |
|127.0.0.1|localhost|Not Allowed|
|localhost|localhost|Allowed|
|localhost|127.0.0.1|Not Allowed|
|Both 127.0.0.1 and localhost|127.0.0.1|Allowed|
|Both 127.0.0.1 and localhost|localhost|Allowed|

I don't think `localhost` and `127.0.0.1` can be interchangeably used. Since, we need maximum cross compatibility, maybe both the `localhost` and `127.0.0.1` should be added to `network_security_config.xml`

**Reply by mhsmith** [reply]
Thanks, I'll look into this. My previous testing was probably on Android 7, 12 or 13.

**Reply by mhsmith** [reply]
I'm seeing the same behavior on an API level 33 emulator, so I've created https://github.com/beeware/briefcase-android-gradle-template/pull/67 to fix it. Thanks for the report.

**Reply by proneon267** [reply]
Thanks for the pull request :)
